### PR TITLE
fix: Handle methodNotAllowed exception for groups

### PR DESCRIPTION
**Pull request by f-alizada**
Closes: #849.
Endpoint for fetching groups (all, linked to product) returns 400 bad request for consumption tier. 
The change introduces possibility to bypass and return empty result even if groups are not allowed, to continue full extraction.

### CODE REVIEW

**Comment by DeagleGross** [comment]
Could you, please, link an issue \ description to know what is the reason of a PR, thanks!

**Comment by f-alizada** [comment]
> Could you, please, link an issue \ description to know what is the reason of a PR, thanks!

Sorry, forgot to update. 
Updated.

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs [hunk 1]
```diff
@@ -30,11 +31,18 @@ public ApiClientBase(IHttpClientFactory httpClientFactory)
...
 
-        protected async Task<string> CallApiManagementAsync(string azToken, string requestUrl, bool useCache = true, ClientHttpMethod method = ClientHttpMethod.GET)
+        /// <summary>
+        ///  Main method to call trigger api management emdpoint.
+        /// </summary>
+        /// <returns>
+        /// string containing response body
+        /// bool indicating if empty default response of the requested resource should be returned
+        /// </returns>
+        protected async Task<(string,bool)> CallApiManagementAsync(string azToken, string requestUrl, bool useCache = true, ClientHttpMethod method = ClientHttpMethod.GET, bool catchMethodNotAllowed = false)
```
Comment (DeagleGross) [inline src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:41]: I still dont get what is the point of `bool` here.

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs [hunk 2]
```diff
@@ -30,11 +31,18 @@ public ApiClientBase(IHttpClientFactory httpClientFactory)
             this.httpClient = httpClientFactory.CreateClient();
         }
 
-        protected async Task<string> CallApiManagementAsync(string azToken, string requestUrl, bool useCache = true, ClientHttpMethod method = ClientHttpMethod.GET)
+        /// <summary>
+        ///  Main method to call trigger api management emdpoint.
```
Comment (DeagleGross) [inline src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:35]: typo "endpoint". nit: I dont think that is a comment, that really tells something useful. It is pretty understandable that this is an API to call apim

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs [hunk 3]
```diff
@@ -49,26 +57,52 @@ protected async Task<string> CallApiManagementAsync(string azToken, string reque
             request.Headers.UserAgent.TryParseAdd($"{Application.Name}/{Application.BuildVersion}");
 
             HttpResponseMessage response = await this.httpClient.SendAsync(request);
-            response.EnsureSuccessStatusCode();
             string responseBody = await response.Content.ReadAsStringAsync();
 
+            try
+            {
+                response.EnsureSuccessStatusCode();
```
Comment (DeagleGross) [inline src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:64]: you have to call `EnsureSuccessStatusCode` before reading content - that is for sure.
Now you:
1) get a unstable behavior in `response.Content.ReadAsStringAsync()`
2) read it for no reason

**Review by DeagleGross — CHANGES_REQUESTED** [review]
IMO your approach is not the best one: I would just add another call in the start of extractor activity to identify sku of apim service - could you please check, if such an endpoint exists?

So that flow would be:
1) extractor process triggered
2) var data = `GET apim/service/data` (unreal endpoint)
3) if service's sku is `consumption`, set it in ExtractorParameters and dont do any calls to unsupported endpoints (like groups)

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs (hunk 3 above)
Comment (f-alizada) [reply src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:64]: agree will move that into handler and later. thanks

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs (hunk 1 above)
Comment (f-alizada) [reply src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:41]: will change the approach since there is a possibility to fetch sku, thanks

**Comment by f-alizada** [comment]
> IMO your approach is not the best one: I would just add another call in the start of extractor activity to identify sku of apim service - could you please check, if such an endpoint exists?
> 
> So that flow would be:
> 
> 1. extractor process triggered
> 2. var data = `GET apim/service/data` (unreal endpoint)
> 3. if service's sku is `consumption`, set it in ExtractorParameters and dont do any calls to unsupported endpoints (like groups)

yes agree, luckly there is such endpoint. thanks for pointing out.

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs (hunk 2 above)
Comment (f-alizada) [reply src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:35]: will update the endpoint type. 
Regarding another nit: Do you want me to delete the description? 
Mainly it was introduced for listing output of the method

**Comment by f-alizada** [comment]
closing the PR since the new approach is completely different.

**Pull request closed without merging**